futures-util/src/async_await/select.rs

```diff
@@ -68,67 +132,412 @@ macro_rules! select {
         // we can ensure that futures aren't polled after completion by use
         // in successive `select!` statements.
         $(
-            $crate::async_await::assert_unpin(&$future_name);
-            $crate::async_await::assert_fused_future(&$future_name);
+            $crate::async_await::assert_unpin(&$fut_name);
+            $crate::async_await::assert_fused_future(&$fut_name);
         )*
 
         #[allow(bad_style)]
-        enum __PrivResult<$($future_name,)*> {
+        enum __PrivResult<$($fut_name,)*> {
             $(
-                $future_name($future_name),
+                $fut_name($fut_name),
             )*
+            __Default,
 			__Complete,
         }
 
-        let __priv_res = await!($crate::future::poll_fn(|lw| {
+        let mut __poll_fn = |lw: &$crate::core_reexport::task::LocalWaker| {
             let mut __any_polled = false;
 
             $(
-                if !$crate::async_await::FusedFuture::is_terminated(& $future_name) {
-                    __any_polled = true;
-                    match $crate::core_reexport::future::Future::poll(
-                        $crate::core_reexport::pin::Pin::new(&mut $future_name), lw)
-                    {
-                        $crate::core_reexport::task::Poll::Ready(x) =>
-                            return $crate::core_reexport::task::Poll::Ready(
-                                __PrivResult::$future_name(x)
-                            ),
-                        $crate::core_reexport::task::Poll::Pending => {},
+                let mut $fut_name = |lw: &_| {
+                    if $crate::async_await::FusedFuture::is_terminated(& $fut_name) {
+                        None
+                    } else {
+                        Some($crate::core_reexport::future::Future::poll(
+                            $crate::core_reexport::pin::Pin::new(&mut $fut_name),
+                            lw,
+                        ).map(__PrivResult::$fut_name))
                     }
-                }
+                };
+                let $fut_name:
+                    &mut dyn FnMut(&$crate::core_reexport::task::LocalWaker)
+                    -> Option<$crate::core_reexport::task::Poll<_>>
+                    = &mut $fut_name;
             )*
+            let mut __select_arr = [$( $fut_name, )*];
+            $crate::rand_reexport::Rng::shuffle(
+                &mut $crate::rand_reexport::thread_rng(),
```

Comment:
This is the only line that requires `std` other than `await!`, correct? What do you think about having an optional argument to specify the `Rng` instance to use, defaulting to `thread_rng`?

Reply:
sounds fine to me! I'd like to leave that for a future change, though, if it's alright with you.

Reply:
Yep, it’s not too useful until we get await! anyway.